snowballDevs/kandoo
# State changes, slideOver rendering, auth bug fix

## Description
This PR deals with the Following:


### Board State
**State Restructuring** : The core board state utilized within the Kanban board has been centralized within the boardContext. Instead of managing individual states scattered across components, this centralisation enables updates to the board state to trigger re-renders in components consuming this context.

**Improved Reusability**: By consuming the context, the Kanban board components can re-render with updated state values, enhancing reusability and maintainability.

**Here is some reading on state structure**:https://react.dev/learn/managing-state#choosing-the-state-structure

---
### Workspace Slideover

**Dedicated Features Folder**: Introduced a dedicated features folder for the Slideover feature. This change separates Slideover-related components making it easier to read our code(imo)

**Breakdown components**: The Slideover has been decomposed into individual components, including Priority Display, Task Description, and Edit Action Buttons. This breakdown enhances code readability and separation of concerns.

**Integration into Kanban Board**: Moved the Workspace Slideover into the Kanban board component.

**Added key prop**: The addition of the key prop to the Workspace Slideover ensures that the state inside the Slideover resets for each distinct task rendered.

**Bug**: There is still a bug with the Slideover transition(on enter). I believe this is caused from adding the key to the slideover and only occurs when selectedTaskId changes. 

---

### Authentication
**Bug Fix**: Addressed an authentication bug related to login. The issue stemmed from an attempt to access the user property of the response object when it was null. This bug has been resolved.

---

### Backend Task Controller
**Change of response**: In response to user task edits in the frontend, the backend now returns the updated task. This allows us to put add the updated task to state.



---
-   [x] Is this related to a specific issue? Is so please specify: #111 #126 

## Type of Changes

Use a check for ✓ the following type of changes:

|     | Type                       |
| --- | -------------------------- |
|  ✓     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|    ✓   | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before

[If applicable, include a picture of the behavior before the changes were made. I.e. a screenshot of the behavior before the changes were made.]

### After
Updated State In the Board Context: Have pulled state upfrom the individual components, and added to the context.
![image](https://github.com/snowballDevs/kandoo/assets/105207353/c50fd23d-f5a5-45ec-9063-27713670b45d)



## Testing Steps / QA Criteria

- Verify that the board state is correctly managed within the boardContext and that changes to the state trigger the re-rendering of components using this context.
- Test the Kanban board's working functionality , such as moving, or editing tasks, to ensure that it works as expected.
- Confirm that the Slideover feature is now organized in a dedicated folder and that all components within it function correctly.
- Check that the Slideover's state is changed when rendering different tasks.

## Checklist:
